### PR TITLE
Build harness in bash

### PR DESCRIPTION
**Pull request by stikonas**
Only tested on m4 for now.

### CODE REVIEW

File: sysa/helpers.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        (
+            build_script="${base_dir}/${2:-${pkg}}.sh"
+            if test -e "${build_script}"; then
+                # shellcheck source=/dev/null
+                . "${build_script}"
+            fi
+            cd "${pkg}" || exit
+
+            call src_prepare
```
Comment (fosslinux) [inline sysa/helpers.sh:32]: Please add messages above each of these to the user, eg `echo "${pkg}: preparing source."`

File: sysa/helpers.sh [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+export PATH=/after/bin
+
+# Common build steps
+# Build function provides a few common stages with default implementation
+# that can be overriden on per package basis.
+build () {
+    pkg=$1
+    cd "$pkg" || exit
```
[the remark this answers is not in the document]
Comment (fosslinux) [reply sysa/helpers.sh:13]: And just below this please add `echo "${pkg}: beginning build"`

File: sysa/helpers.sh [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                . "${build_script}"
+            fi
+            cd "${pkg}" || exit
+
+            call src_prepare
+            call src_configure
+            call src_build
+            call src_install
+        )
+    )
```
Comment (fosslinux) [inline sysa/helpers.sh:37]: Below, `echo "${pkg}: build successful"`.

**Review by fosslinux — CHANGES_REQUESTED** [review]
I approve of the lack of `-x`.

This is overall a pretty good implementation. I have pretty strong opinions about how things like these helper.sh 's should work after seeing bad ones in distros.

My main complaint is the lack of comments, please add more comments explaining things less than obvious (and even obvious things), particular the test for function thing.

I have a few other things I would like to add but they can be done later.

File: sysa/helpers.sh [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    pkg=$1
+    cd "$pkg" || exit
+    base_dir="${PWD}"
+    patch_dir="${base_dir}/patches"
+    mk_dir="${base_dir}/mk"
+
+    rm -rf "build"
+    mkdir -p "build"
+    (
+        cd "build" || exit
```
[the remark this answers is not in the document]
Comment (stikonas) [reply sysa/helpers.sh:21]: This one is mainly to silience the shellcheck. We can add silencing comment instead but that seemed longer anyway.

[the remark this answers is not in the document]
Comment (fosslinux) [reply sysa/helpers.sh:21]: I dislike adding complexity to code to silence linters.

[the remark this answers is not in the document]
Comment (stikonas) [reply sysa/helpers.sh:21]: Removed exit and added `-e` to shebang silence linter. It's sourced file anyway.

File: sysa/helpers.sh [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    elif test -e "${bzip2_source}"; then
+        tar -xjf "${bzip2_source}"
+    elif test -e "${xz_source}"; then
+        tar -xJf "${xz_source}"
+    fi
+}
+
+default_src_prepare() {
+    for p in "${patch_dir}"/*.patch; do
+        test -e "${p}" || continue
```
[the remark this answers is not in the document]
Comment (stikonas) [reply sysa/helpers.sh:56]: ok, I'll do `test -d`. That will imply not allowing empty patch dir though.

File: sysa/helpers.sh [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+build () {
+    pkg=$1
+    cd "$pkg" || exit
+    base_dir="${PWD}"
+    patch_dir="${base_dir}/patches"
+    mk_dir="${base_dir}/mk"
+
+    rm -rf "build"
+    mkdir -p "build"
+    (
```
[the remark this answers is not in the document]
Comment (stikonas) [reply sysa/helpers.sh:23]: We can replace it with `cd ..` at the end if it's more readable.

[the remark this answers is not in the document]
Comment (fosslinux) [reply sysa/helpers.sh:31]: Oooh, I see, for the cd; well, please just comment above subshell to perform cd or something like that.

**Review by fosslinux — CHANGES_REQUESTED** [review]
Still would also like error messages on those `|| exit`'s.

**Comment by stikonas** [comment]
There is still an issue with subshells that needs to be resolved before merge.

**Comment by stikonas** [comment]
Subshells got removed because trap from subshells doesn't properly work in our bootstrapped bash (same test code works fine in modern bash).

**Review by fosslinux — APPROVED** [review]

**Pull request merged**